### ISSUE TITLE
High-level version of lax.scatter variants for more intuitive use of scatter

### ISSUE THREAD

**Issue by AranKomat** [google/jax#8487]
`lax.gather` has its high-level counterpart as `jax.numpy.take_along_axis`, which is similar to `torch.gather`. However, for scatter operations there is no such counterpart, and one has to use `lax.scatter` variants, which require understanding of semantics of XLA's scatter: https://www.tensorflow.org/xla/operation_semantics#scatter. 

To this day, I still don't understand the semantics, and I think so are many users. When they want to translate PyTorch to JAX, this may pose a problem given the ubiquity of scatter operations. I wrote a high-level scatter similar to PyTorch using nested vmaps for example (not sure if this implementation is the most straightforward one ... if not, plz lmk). I think it'd be great to add something like this to `jax.numpy`. 

 ```python
 import jax
import jax.numpy as jnp
from functools import partial
from jax import vmap

def scatter(input, dim, index, src, reduce=None):
    # Works like PyTorch's scatter. See https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html
    
    dnums = jax.lax.ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
    
    if reduce is None:
        _scatter = jax.lax.scatter
    elif reduce == "add":
        _scatter = jax.lax.scatter_add
    elif reduce == "multiply":
        _scatter = jax.lax.scatter_mul
        
    _scatter = partial(_scatter, dimension_numbers=dnums)
    vmap_inner = partial(vmap, in_axes=(0, 0, 0), out_axes=0)
    vmap_outer = partial(vmap, in_axes=(1, 1, 1), out_axes=1)

    for idx in range(len(input.shape)):
        if idx == dim:
            pass
        elif idx < dim:
            _scatter = vmap_inner(_scatter)
        else:
            _scatter = vmap_outer(_scatter)
            
    return _scatter(input, jnp.expand_dims(index, axis=-1), src)

index = jnp.asarray([[0, 1], [1, 0]])
src = jnp.asarray([[1, 2], [3, 4]])
input = jnp.zeros((2, 2)).astype(jnp.int32)

output = scatter(input, 0, index, src)
print(output)
 ```

**Comment by jakevdp** [comment]
Thanks for the comment. For what it's worth, JAX does offer a number of high-level wrappers around XLA's `scatter`, which are discussed in https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html#jax.numpy.ndarray.at

I think the root issue here is that what is meant by "scatter" in pytorch is quite different than what is meant by "scatter" in XLA.

**Comment by jakevdp** [comment]
I was looking at it; If I'm understanding the function, I think a simpler implementation of pytorch's `scatter_` using these higher-level wrappers might look something like this:
```python
def scatter2(input, dim, index, src, reduce=None):
  assert index.shape == input.shape == src.shape
  idx = jnp.meshgrid(*(jnp.arange(n) for n in index.shape))
  idx = tuple([*idx[:dim], index, *idx[dim + 1:]])
  return getattr(input.at[idx], reduce or "set")(src)
```

**Comment by AranKomat** [comment]
Thank you very much for your response! Admittedly, I didn't know that there's `ndarray.at` option, and your solution is very elegant! However, I tried your scatter on the above example and got a result slightly different from PyTorch's output. With a different example, I got "size incompatible" error. Maybe there was an error somewhere.

**Comment by jakevdp** [comment]
Yeah - I'm not sure I totally understand what pytorch's scatter is meant to do (particularly when input shapes don't match), but I suspect it can be implemented in terms of XLA's scatter using an approach similar to the one in my comment.

**Comment by guozhiyu** [comment]
@AranKomat There is still no implementation of PyTorch scatter in JAX. I have tested `def scatter(input, dim, index, src, reduce=None):
  idx = jnp.meshgrid(*(jnp.arange(n) for n in input.shape), sparse=True)
  idx[dim] = index
  return getattr(input.at[tuple(idx)], reduce or "set")(src)`, it outputs different results with PyTorch. Can you provide information about how you solve this finally?

**Comment by hawkinsp** [comment]
@guozhiyu It's impossible to say what you should do without a minimum reproducible example. Can you share concrete inputs and outputs in both JAX and pytorch that differ?

**Comment by guozhiyu** [comment]
Thank you for your reply! I tested the example Aran provided: 
```
index = jnp.asarray([[0, 1], [1, 0]])
src = jnp.asarray([[1, 2], [3, 4]])
input = jnp.zeros((2, 2)).astype(jnp.int32)

output = scatter(input, 0, index, src, reduce='add')
print(output)
```
 output is 
```
[[1 4]
 [2 3]]
```
, but in PyTorch, 
```
index = torch.tensor([[0, 1], [1, 0]])
src = torch.tensor([[1, 2], [3, 4]],dtype=torch.int32)
input = torch.zeros((2, 2),dtype=torch.int32)
output = input.scatter_(0, index, src, reduce="add")
```
 the output is 
```
[[1, 4],
[3, 2]]
```

**Comment by jakevdp** [comment]
For what it's worth, I never intended to imply that the four-line `scatter` function in my comment above was a full-featured semantic equivalent of pytorch's `scatter`, so it's not surprising that it produces different results in some cases. I still don't fully understand the indended semantics of pytorch's `scatter`... For what it's worth my suggestion would be to re-express your intended computation in terms of JAX's native primitives, rather than trying to replicate pytorch's native primitives in JAX.

From your example inputs, perhaps this is what you're looking for?
```python
import jax
import jax.numpy as jnp

index = jnp.asarray([[0, 1], [1, 0]])
src = jnp.asarray([[1, 2], [3, 4]])

out = jax.vmap(lambda x, i: x[i], in_axes=1, out_axes=1)(src, index)
print(out)
# [[1 4]
#  [3 2]]
```

**Comment by AranKomat** [comment]
@guozhiyu Hi. The post was made such a long time ago, so I can't guarantee if my implementation of PyTorch scatter works or not, but this is what I've found from my old repo:

```python
def scatter(input, dim, index, src, reduce=None):
    # JAX-port of PyTorch's scatter. See https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html
    # One can simplify this implementation with ndarray.at (ref: https://github.com/google/jax/issues/8487)

    dnums = jax.lax.ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
    
    if reduce is None:
        _scatter = jax.lax.scatter
    elif reduce == "add":
        _scatter = jax.lax.scatter_add
    elif reduce == "multiply":
        _scatter = jax.lax.scatter_mul
        
    _scatter = partial(_scatter, dimension_numbers=dnums)
    vmap_inner = partial(vmap, in_axes=(0, 0, 0), out_axes=0)

    for _ in range(len(input.shape)-1):
        _scatter = vmap_inner(_scatter)
    swap = lambda x: jnp.swapaxes(x, dim, -1)
    input, index, src = list(map(swap, (input, index, src)))
    return swap(_scatter(input, jnp.expand_dims(index, axis=-1), src))
 ```

**Comment by guozhiyu** [comment]
@AranKomat Your code works well. Thank you very much!
